### ISSUE TITLE
Fix flakey TestGetProfiles

### ISSUE THREAD

**Issue by siggy** [linkerd/linkerd2#3332]
in travis:
https://travis-ci.org/linkerd/linkerd2/jobs/577582791#L749-L753

locally:
```bash
$ while go test -count 1 github.com/linkerd/linkerd2/controller/api/destination; do ls > /dev/null ; done
ok  	github.com/linkerd/linkerd2/controller/api/destination	0.586s
...
ok  	github.com/linkerd/linkerd2/controller/api/destination	0.602s
time="2019-08-27T16:06:13-07:00" level=error msg="A request match must have a field set" test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=error msg="A response match must have a field set" test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=error msg="Range maximum cannot be smaller than minimum" test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="waiting for caches to sync"
time="2019-08-27T16:06:13-07:00" level=info msg="caches synced"
time="2019-08-27T16:06:13-07:00" level=info msg="waiting for caches to sync"
time="2019-08-27T16:06:13-07:00" level=info msg="caches synced"
time="2019-08-27T16:06:13-07:00" level=info msg="Establishing watch on endpoint [ns/name1:8989]" component=endpoints-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=warning msg="failed to retrieve replicaset from k8s ns/rs-1: replicaset.apps \"rs-1\" not found"
time="2019-08-27T16:06:13-07:00" level=info msg="Stopping watch on endpoint [ns/name1:8989]" component=endpoints-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="waiting for caches to sync"
time="2019-08-27T16:06:13-07:00" level=info msg="caches synced"
time="2019-08-27T16:06:13-07:00" level=info msg="waiting for caches to sync"
time="2019-08-27T16:06:13-07:00" level=info msg="caches synced"
time="2019-08-27T16:06:13-07:00" level=info msg="Establishing watch on service ns/name1" component=traffic-split-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="Establishing watch on profile other/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="Establishing watch on profile ns/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="Stopping watch on profile ns/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="Stopping watch on profile other/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="Stopping watch on profile ns/name1" component=traffic-split-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="waiting for caches to sync"
time="2019-08-27T16:06:13-07:00" level=info msg="caches synced"
time="2019-08-27T16:06:13-07:00" level=info msg="Establishing watch on service ns/name1" component=traffic-split-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="Establishing watch on profile client-ns/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="Establishing watch on profile ns/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="Stopping watch on profile ns/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="Stopping watch on profile client-ns/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1ec8140
time="2019-08-27T16:06:13-07:00" level=info msg="Stopping watch on profile ns/name1" component=traffic-split-watcher test=0x1ec8140
--- FAIL: TestGetProfiles (0.33s)
    --- FAIL: TestGetProfiles/Returns_server_profile (0.11s)
        server_test.go:228: Expected 1 or 2 updates but got 3: [retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > > ]
FAIL
FAIL	github.com/linkerd/linkerd2/controller/api/destination	0.624s
```

## Stack traces

In the failing case, the 3rd update comes from a different code path than the first two:

### First update

```
SEND 1: [retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > > ]
goroutine 28 [running]:
runtime/debug.Stack(0x57, 0x0, 0x0)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x22
github.com/linkerd/linkerd2/controller/api/destination.(*bufferingGetProfileStream).Send(0xc0003da2a0, 0x31b2220, 0x102ec61, 0xc0000317f8)
	/Users/sig/code/linkerd2/controller/api/destination/server_test.go:143 +0x132
github.com/linkerd/linkerd2/controller/api/destination.(*profileTranslator).Update(0xc0003e44e0, 0x0)
	/Users/sig/code/linkerd2/controller/api/destination/profile_translator.go:49 +0x14e
github.com/linkerd/linkerd2/controller/api/destination.(*trafficSplitAdaptor).publish(0xc000438ae0)
	/Users/sig/code/linkerd2/controller/api/destination/traffic_split_adaptor.go:70 +0x439
github.com/linkerd/linkerd2/controller/api/destination.(*trafficSplitAdaptor).Update(0xc000438ae0, 0x0)
	/Users/sig/code/linkerd2/controller/api/destination/traffic_split_adaptor.go:39 +0x3f
github.com/linkerd/linkerd2/controller/api/destination.(*primaryProfileListener).Update(0xc00059bc30, 0x0)
	/Users/sig/code/linkerd2/controller/api/destination/fallback_profile_listener.go:72 +0xb2
github.com/linkerd/linkerd2/controller/api/destination/watcher.(*profilePublisher).subscribe(0xc000438cc0, 0x24998e0, 0xc00059bc30)
	/Users/sig/code/linkerd2/controller/api/destination/watcher/profile_watcher.go:175 +0xd0
github.com/linkerd/linkerd2/controller/api/destination/watcher.(*ProfileWatcher).Subscribe(0xc0003e8480, 0x2271211, 0x5, 0xc0003d0160, 0x1c, 0x24998e0, 0xc00059bc30, 0x5, 0xc0003d0160)
	/Users/sig/code/linkerd2/controller/api/destination/watcher/profile_watcher.go:78 +0x141
github.com/linkerd/linkerd2/controller/api/destination.(*server).GetProfile(0xc000438a80, 0xc0001806e0, 0x24ec8a0, 0xc0003da2a0, 0x0, 0x0)
	/Users/sig/code/linkerd2/controller/api/destination/server.go:203 +0xca4
github.com/linkerd/linkerd2/controller/api/destination.TestGetProfiles.func1(0xc000308100)
	/Users/sig/code/linkerd2/controller/api/destination/server_test.go:221 +0x19a
testing.tRunner(0xc000308100, 0x2306b30)
	/usr/local/go/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:916 +0x35a
time="2019-09-03T17:47:45-07:00" level=info msg="Establishing watch on profile ns/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1f22d80
```

### Second update

Identical to first update except for `backupProfileListener`:

```
SEND 2: [retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > > ]
goroutine 28 [running]:
runtime/debug.Stack(0x115, 0x0, 0x0)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x22
github.com/linkerd/linkerd2/controller/api/destination.(*bufferingGetProfileStream).Send(0xc0003da2a0, 0xc00063e180, 0x1b, 0xc0000317e8)
	/Users/sig/code/linkerd2/controller/api/destination/server_test.go:143 +0x132
github.com/linkerd/linkerd2/controller/api/destination.(*profileTranslator).Update(0xc0003e44e0, 0xc0002ac000)
	/Users/sig/code/linkerd2/controller/api/destination/profile_translator.go:58 +0x122
github.com/linkerd/linkerd2/controller/api/destination.(*trafficSplitAdaptor).publish(0xc000438ae0)
	/Users/sig/code/linkerd2/controller/api/destination/traffic_split_adaptor.go:72 +0x408
github.com/linkerd/linkerd2/controller/api/destination.(*trafficSplitAdaptor).Update(0xc000438ae0, 0xc00029ca00)
	/Users/sig/code/linkerd2/controller/api/destination/traffic_split_adaptor.go:39 +0x3f
github.com/linkerd/linkerd2/controller/api/destination.(*backupProfileListener).Update(0xc00059bc40, 0xc00029ca00)
	/Users/sig/code/linkerd2/controller/api/destination/fallback_profile_listener.go:93 +0xb5
github.com/linkerd/linkerd2/controller/api/destination/watcher.(*profilePublisher).subscribe(0xc0000e7200, 0x24998a0, 0xc00059bc40)
	/Users/sig/code/linkerd2/controller/api/destination/watcher/profile_watcher.go:175 +0xd0
github.com/linkerd/linkerd2/controller/api/destination/watcher.(*ProfileWatcher).Subscribe(0xc0003e8480, 0x228f010, 0x2, 0xc0003d0320, 0x1c, 0x24998a0, 0xc00059bc40, 0x2, 0xc0003d0320)
	/Users/sig/code/linkerd2/controller/api/destination/watcher/profile_watcher.go:78 +0x141
github.com/linkerd/linkerd2/controller/api/destination.(*server).GetProfile(0xc000438a80, 0xc0001806e0, 0x24ec8a0, 0xc0003da2a0, 0x0, 0x0)
	/Users/sig/code/linkerd2/controller/api/destination/server.go:216 +0x106a
github.com/linkerd/linkerd2/controller/api/destination.TestGetProfiles.func1(0xc000308100)
	/Users/sig/code/linkerd2/controller/api/destination/server_test.go:221 +0x19a
testing.tRunner(0xc000308100, 0x2306b30)
	/usr/local/go/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:916 +0x35a
time="2019-09-03T17:47:46-07:00" level=info msg="Stopping watch on profile ns/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1f22d80
```

### Third update (causes failure)

Note this originates from `k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)`

```
SEND 3: [retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > > ]
goroutine 126 [running]:
runtime/debug.Stack(0x1d3, 0x0, 0x0)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x22
github.com/linkerd/linkerd2/controller/api/destination.(*bufferingGetProfileStream).Send(0xc000410660, 0xc000324ba0, 0x1b, 0xc00061ca28)
	/Users/sig/code/linkerd2/controller/api/destination/server_test.go:143 +0x132
github.com/linkerd/linkerd2/controller/api/destination.(*profileTranslator).Update(0xc000497320, 0xc0004a6a00)
	/Users/sig/code/linkerd2/controller/api/destination/profile_translator.go:58 +0x122
github.com/linkerd/linkerd2/controller/api/destination.(*trafficSplitAdaptor).publish(0xc000324a80)
	/Users/sig/code/linkerd2/controller/api/destination/traffic_split_adaptor.go:72 +0x408
github.com/linkerd/linkerd2/controller/api/destination.(*trafficSplitAdaptor).Update(0xc000324a80, 0xc000196500)
	/Users/sig/code/linkerd2/controller/api/destination/traffic_split_adaptor.go:39 +0x3f
github.com/linkerd/linkerd2/controller/api/destination.(*backupProfileListener).Update(0xc0002e7af0, 0xc000196500)
	/Users/sig/code/linkerd2/controller/api/destination/fallback_profile_listener.go:93 +0xb5
github.com/linkerd/linkerd2/controller/api/destination/watcher.(*profilePublisher).update(0xc00035cd20, 0xc000196500)
	/Users/sig/code/linkerd2/controller/api/destination/watcher/profile_watcher.go:207 +0x125
github.com/linkerd/linkerd2/controller/api/destination/watcher.(*ProfileWatcher).addProfile(0xc00003cc80, 0x2241a40, 0xc000196500)
	/Users/sig/code/linkerd2/controller/api/destination/watcher/profile_watcher.go:103 +0x86
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
	/Users/sig/code/go/pkg/mod/k8s.io/client-go@v0.0.0-20181213151034-8d9ed539ba31/tools/cache/controller.go:195
k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x0, 0x0, 0x0)
	/Users/sig/code/go/pkg/mod/k8s.io/client-go@v0.0.0-20181213151034-8d9ed539ba31/tools/cache/shared_informer.go:554 +0x26d
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0xc000471e38, 0x102e3bf, 0xc0002d7480)
	/Users/sig/code/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127105237-2b1284ed4c93/pkg/util/wait/wait.go:203 +0xde
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/Users/sig/code/go/pkg/mod/k8s.io/client-go@v0.0.0-20181213151034-8d9ed539ba31/tools/cache/shared_informer.go:548 +0x89
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000613768)
	/Users/sig/code/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127105237-2b1284ed4c93/pkg/util/wait/wait.go:133 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000471f68, 0xdf8475800, 0x0, 0x1, 0xc0000c6720)
	/Users/sig/code/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127105237-2b1284ed4c93/pkg/util/wait/wait.go:134 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/Users/sig/code/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127105237-2b1284ed4c93/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*processorListener).run(0xc00036a880)
	/Users/sig/code/go/pkg/mod/k8s.io/client-go@v0.0.0-20181213151034-8d9ed539ba31/tools/cache/shared_informer.go:546 +0x9c
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc000166fb0, 0xc0004c0760)
	/Users/sig/code/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127105237-2b1284ed4c93/pkg/util/wait/wait.go:71 +0x4f
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/Users/sig/code/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127105237-2b1284ed4c93/pkg/util/wait/wait.go:69 +0x62
time="2019-09-03T17:47:53-07:00" level=info msg="Stopping watch on profile ns/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1f22d80
time="2019-09-03T17:47:53-07:00" level=info msg="Stopping watch on profile other/name1.ns.svc.mycluster.local" component=profile-watcher test=0x1f22d80
time="2019-09-03T17:47:53-07:00" level=info msg="Stopping watch on profile ns/name1" component=traffic-split-watcher test=0x1f22d80
--- FAIL: TestGetProfiles (1.12s)
    --- FAIL: TestGetProfiles/Returns_server_profile (1.12s)
        server_test.go:237: Got 3: [retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > > ]
        server_test.go:239: Expected 1 or 2 updates but got 3: [retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > > ]
```

**Comment by alpeb** [comment]
Thanks @siggy for the stack traces :+1: 

TL;DR;
Service profiles are added and the listeners are immediately set up. Usually the client-go notification is triggered before the listeners are setup, but sometimes (very rarely) it gets delayed a bit. Unlikely to happen in production.

---
As shown in there, the first two updates come from when the primary and secondary listeners are set up, when `GetProfile()` is called:
https://github.com/linkerd/linkerd2/blob/981f5bc85dd84aa02524c3cd822bdd9a2c1c0756/controller/api/destination/profile_translator.go#L49

One can receive the third update sporadically; sometimes it takes just a dozen runs to receive, sometimes as much as a couple thousands!
As shown in the third stack trace above, that update comes from client-go, when it detects the addition of a new resource.
Inside `makeServer()` the resources are added, and then the watchers are set up. By the time the watchers are set up, the resources have already been (fakely) persisted and so the watchers aren't notified. Except for those rare cases when those notifications take more time than usual, in which case we get the third update.

When a resource is added, the notifications are ultimately scheduled using client-go's `wait.Start()` function, to which I added some debug statements:
```go
func (g* Group) Start(f func()) {
  g.wg.Add(1)
  fmt.Println("wait.Start()")
  go func() {
    fmt.Println("wait.Start().f()")
    defer g.wg.Done()
    f()
  }()
}
```

When the test fails we get something like:
```
time="2020-01-15T07:45:25-05:00" level=info msg="waiting for caches to sync"
wait.Start()
wait.Start()
wait.Start()
wait.Start()
wait.Start()
wait.Start()
wait.Start()
wait.Start()
wait.Start().f()
.. a bunch more of these
time="2020-01-15T07:45:25-05:00" level=info msg="caches synced"
wait.Start()
wait.Start()
wait.Start().f()
 ..more
time="2020-01-15T07:45:25-05:00" level=debug msg="GetProfile(scheme:\"k8s\" path:\"name1.ns.svc.mycluster.local:8989\" context_token:\"ns:other\" )" test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=info msg="Establishing watch on service ns/name1" component=traffic-split-watcher test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=info msg="Establishing watch on profile other/name1.ns.svc.mycluster.local" component=profile-watcher test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=info msg="Establishing watch on profile ns/name1.ns.svc.mycluster.local" component=profile-watcher test=TestGetProfiles/Returns_server_profile
wait.Start().f()
wait.Start().f()
wait.Start().f()
wait.Start().f()
wait.Start().f()
wait.Start().f()
wait.Start().f()
time="2020-01-15T07:45:25-05:00" level=debug msg="Updating service for ns/name1" component=service-publisher ns=ns svc=name1 test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=error msg="Index with name ip does not exist" component=ip-watcher test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=error msg="Index with name ip does not exist" component=ip-watcher test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=debug msg="Sending profile update: routes:<condition:<path:<regex:\"/a/b/c\" > > metrics_labels:<key:\"route\" value:\"route1\" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > > " component=profile-translator test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=debug msg="GetProfile(scheme:\"k8s\" path:\"name1.ns.svc.mycluster.local:8989\" context_token:\"ns:other\" ) cancelled" test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=info msg="Stopping watch on profile ns/name1.ns.svc.mycluster.local" component=profile-watcher test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=debug msg="Updating endpoints for ns/name1" component=service-publisher ns=ns svc=name1 test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=debug msg="Updating profile" component=profile-publisher ns=ns profile=name1.ns.svc.mycluster.local test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=debug msg="Sending profile update: routes:<condition:<path:<regex:\"/a/b/c\" > > metrics_labels:<key:\"route\" value:\"route1\" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > > " component=profile-translator test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=debug msg="Updating profile" component=profile-publisher ns=client-ns profile=name1.ns.svc.mycluster.local test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=info msg="Stopping watch on profile other/name1.ns.svc.mycluster.local" component=profile-watcher test=TestGetProfiles/Returns_server_profile
time="2020-01-15T07:45:25-05:00" level=info msg="Stopping watch on profile ns/name1" component=traffic-split-watcher test=TestGetProfiles/Returns_server_profile
--- FAIL: TestGetProfiles (0.11s)
    --- FAIL: TestGetProfiles/Returns_server_profile (0.11s)
        server_test.go:232: Expected 1 or 2 updates but got 3: [retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > >  routes:<condition:<path:<regex:"/a/b/c" > > metrics_labels:<key:"route" value:"route1" > timeout:<seconds:10 > > retry_budget:<retry_ratio:0.2 min_retries_per_second:10 ttl:<seconds:10 > > ]
FAIL
FAIL    github.com/linkerd/linkerd2/controller/api/destination  0.149s
FAIL
```
We see there's a bunch of `wait.Start().f()` after the call to `GetProfile()` which means the goroutine in `Start()` got delayed (GC run maybe?).

This is a very rare scenario that I don't think we'd see in prod: adding a service profile and a couple of millisecs later have the destination service set up its listeners. So I think it's safe to re-enable this test and just set the expected stream updated from 1 to 3.

**Comment by siggy** [comment]
Thanks for the thorough analysis @alpeb. General approach sounds good. Naive question: do we lose some fidelity in setting the update count from 1 to 3? If so, is there another way to validate what the test was previously trying to validate?

**Comment by alpeb** [comment]
Hey @siggy sorry for the late reply...
I tested adding a 1s sleep after the setup, and that seems to remove the late incoming update, so we can leave the test as it is. Does that sound good?

**Comment by siggy** [comment]
Hey @alpeb. Hmm, I think your original plan of a range from 1 to 3 sounds better than adding a sleep. Let's go with that. I think the primary goal of a test like this is to ensure _something_ is happening.
